1259 multi select fixes

herronelou opened the pull request:
<!--
Thanks for creating this pull request 🤗

Please make sure that the pull request is limited to one type (docs, feature, etc.) and keep it as small as possible. You can open multiple prs instead of opening a huge one.
-->

<!-- If this pull request closes an issue, please mention the issue number below -->

Closes #1259 

## 📑 Description

<!-- Add a brief description of the PR -->

## Status

- [ ] Not Completed
- [x] Completed

## ✅ Checks

<!-- Make sure your PR passes the tests and do check the following fields as needed - -->

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] I have checked the page with https://validator.unl.edu/
- [x] All the tests have passed
- [x] My pull request is based on the latest commit (not the npm version).

<!--

Sync fork from GitHub dropdown and update your local branch.

```sh
git pull origin main
git checkout your-branch
git rebase main
npm run test
git push
```

Then create a PR from your GitHub repo.

Or if you are using the command line:

```sh
git checkout main
git fetch upstream // I'm assuming you set the upstream
git merge upstream/main
```

Then change to your branch:

```sh
git checkout my-new-feature
git merge main
```

If you may need to resolve merge conficts if your local branch had unique commits.

Now you are ready to submit your PR.

It’s a good idea to sync from time to
time, so you aren’t left too far behind the parent branch.

-->

## ℹ Additional Information

<!-- Any additional information like breaking changes, dependencies added, screenshots, comparisons between new and old behavior, etc. -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
    - Enhanced the MultiSelect component with keyboard navigation. Users can now use arrow keys to navigate, Enter or Space to select, and Escape to close the dropdown.
    - Improved visibility and interaction of the MultiSelect component with new styling for the active item.
- **Bug Fixes**
    - Adjusted the focus trap utility to correctly skip elements that are hidden, ensuring better accessibility and user experience.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

herronelou commented:
@shinokada are you the one who reviews PRs?

shinokada commented:
Is it possible to create an example that you solved?
Your example and explanation can be at https://flowbite-svelte.com/docs/forms/select#MultiSelect .

herronelou commented:
Hello.
What kind of examples are you looking for?
I have posted an example in the linked issue #1259


On Thu, Feb 29, 2024, 23:34 Shinichi Okada ***@***.***> wrote:

> Is it possible to create an example that you solved?
> Your example and explanation can be at
> https://flowbite-svelte.com/docs/forms/select#MultiSelect .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/themesberg/flowbite-svelte/pull/1260#issuecomment-1972671158>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAV5HA5BDDNOFKHSQXYDTYWAVRHAVCNFSM6AAAAABD3IDGKWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZSGY3TCMJVHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

shinokada commented:
Yes, that will be fine. Can you add it to the docs?

herronelou commented:
I can, I'm just not too sure what to add. Every other component tends to
have keyboard navigation by default based on the original HTML elements,
this is just trying to mimic it.

Should I add a short mention of the keys that have been implemented?

On Fri, Mar 1, 2024, 06:53 Shinichi Okada ***@***.***> wrote:

> Yes, that will be fine. Can you add it to the docs?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/themesberg/flowbite-svelte/pull/1260#issuecomment-1973339307>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJAV5BLJNWMSXDE6OOMZJLYWCI6TAVCNFSM6AAAAABD3IDGKWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZTGMZTSMZQG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

herronelou commented:
@shinokada I've added a little section in the docs, I'm not sure if this is what you had in mind?

shinokada commented:
You wrote, "then press enter or spacebar to toggle the highlighted item.". But enter or spacebar don't work to select items.

herronelou commented:
Hi @shinokada, it works for me here, can you give me more information about what doesn't work for you? 
Upon highlighting an element with the up/down arrows, pressing space or enter should toggle that element.

shinokada commented:
Go to https://flowbite-svelte-lca18iqht-themesberg.vercel.app/docs/forms/select#Basic_example and use arrow down/up and hit the return key. It doesn't select an item. All MultiSelect examples do not work with return or space key. 

I tested with Chrome and FF.

herronelou commented:
It doesn't work for me either on the link above, nor on my local build in the docs, but it works with my local build in my real project. I'll try to see if I can make it work in the docs context.

```
  System:
    OS: Windows 11 10.0.22631
    CPU: (24) x64 12th Gen Intel(R) Core(TM) i9-12900
    Memory: 31.99 GB / 63.73 GB
  Binaries:
    Node: 18.15.0 - ~\AppData\Roaming\JetBrains\WebStorm2023.3\node\node-v18.15.0-win-x64\node.EXE
    npm: 9.5.0 - ~\AppData\Roaming\JetBrains\WebStorm2023.3\node\node-v18.15.0-win-x64\npm.CMD
    pnpm: 8.15.6 - ~\AppData\Roaming\JetBrains\WebStorm2023.3\node\node-v18.15.0-win-x64\pnpm.CMD
  Browsers:
    Edge: Chromium (123.0.2420.97)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @sveltejs/kit: ^2.5.1 => 2.5.1
    svelte: ^4.2.11 => 4.2.11
    vite: ^5.1.4 => 5.1.4
```

herronelou commented on src/lib/forms/MultiSelect.svelte:
```diff
@@ -66,6 +71,40 @@
+    }
+    else if (event.key === `Enter` || event.key === ` `) {
+      // If the focused element is our div, prevent the form submission and show the dropdown
+      if (focusedElement?.role === 'listbox') {
+        event.preventDefault(); // prevent enter key from triggering form submission
+        // Show the dropdown when pressing enter/space while in focus.
```
@shinokada This check is preventing it from working in the context of the documentation, as the focusedElement is 
```
<iframe class="w-full h-full" title="iframe-code-content"></iframe>
```
I don't think this was strictly necessary but I was trying to be conservative.
I can remove the check, which would simplify the function, and make it work in the docs.
Again I'm normally a python dev so all this is quite new to me, I'd love your advice on the matter.

herronelou commented:
@shinokada I have rebased this on the latest main branch, also cleaned up the commits into 3 clearly defined commits, while taking into account the comments from code rabbit.

Let me know if anything else is needed.

shinokada commented:
Thank you for your contribution.
